File: rcc/communications/client/file/abstract_file_client.py
from abc import ABC

from framework.logger import Logger
from rcc.utils.shell import SHELL


class AbstractFileClient(ABC):
    """Abstract class for file client"""

    def __init__(self, path: str):
        self.__path = path
        self._logger = Logger(self.__class__.__name__)
        self._available = SHELL.run_command(f"ls {self.__path}", check=False) == 0

    @property
    def available(self):
        """Availability flag"""
        return self._available

    def read(self, head=0, tail=0):
        """Read content from file with head and tail optional"""
        cmd = f"cat {self.__path}"
        if head > 0:
            cmd += f" | head -n{head}"
        if tail > 0:
            cmd += f" | tail -n{tail}"

        return SHELL.run_command(cmd)[1]

    def write(self, content, sudo=False):
        """Write content to file"""
        if sudo:
            SHELL.run_sudo_command(f"echo '{content}' | tee {self.__path}")

        SHELL.run_command(f"echo '{content}' | tee {self.__path}")

File: Framework/src/clients/abstract/abstract_file_client.cpp

#include "framework/clients/abstract/abstract_file_client.hpp"

#include "framework/utils/file_utils.hpp"

std::string AbstractFileClient::read(int head, int tail) {
	if (!available_) {
		throw std::runtime_error("File " + path_ + " doesn't exist");
	}

	std::string cmd = "cat " + path_;
	if (head > 0) {
		cmd += " | head -n" + std::to_string(head);
	}
	if (tail > 0) {
		cmd += " | tail -n" + std::to_string(tail);
	}

	if (sudo_) {
		return shell.run_elevated_command(cmd).stdout_str;
	} else {
		return shell.run_command(cmd).stdout_str;
	}
}

void AbstractFileClient::write(const std::string& content) {
	std::string cmd = "echo '" + content + "' | tee " + path_;
	if (sudo_) {
		shell.run_elevated_command(cmd);
	} else {
		shell.run_command(cmd);
	}
}

bool AbstractFileClient::available() {
	return available_;
}

AbstractFileClient::AbstractFileClient(const std::string& path, const std::string& name, bool sudo, bool required)
	: Loggable(name), path_(path), sudo_(sudo) {
	available_ = FileUtils::exists(path);
	if (!available_ && required) {
		throw std::runtime_error("File " + path_ + " doesn't exist");
	}
}
